### DISCUSSION TITLE
The results of parallel and serial calculations in MOOSE are different in the thermo-mechanical coupling calculation

### DISCUSSION THREAD

**Discussion by js-jixu** [idaholab/moose#21878] [Q&A Modules: Solid mechanics] [answered]
Hi, everyone.

I want to do some thermomechanical coupling problems with MOOSE. and I want to set the model to have no displacement in the Z direction, so I set the boundary condition of displacements like this：

```
  [./top_and_bottom]
    type = DirichletBC
    variable = disp_z
    boundary = 'outer_bottom outer_top inner_bottom inner_top'
    value = 0
  [../]
```

There are two problems. The first one is that when I calculate with different numbers of cores, I get different results. For example, when I use one core, 4 cores and 8 cores for calculation, the results obtained are very different. It seems that the serial computation results is reasonable. The second one is that the distributions of disp_x and disp_y are not symmetrical to the x and y axes. But they are supposed to be symmetrical. If I want to make disp_x and disp_y symmetrical, should I set some boundary conditions for them?

The calculation results of using 1 core, 4 cores, and 8 cores are as follows:

![1core](https://user-images.githubusercontent.com/99813361/185423113-f0cc07ee-3608-4788-8175-c981afa194c1.png)
![4mpi](https://user-images.githubusercontent.com/99813361/185423148-f0f68921-6b00-4b5f-ae21-2494b513f699.png)
![8mpi](https://user-images.githubusercontent.com/99813361/185424157-33d7befb-bbea-4f17-a0b7-b9b22257aa6d.png)

There are output files and mesh file.
[attached files.zip](https://github.com/idaholab/moose/files/9374867/attached.files.zip)

There is complete input file:
```
ini_temp = 593 # K

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
  file = '2layers_3d_4parts.msh'
[]

[Variables]
  [./T]
    initial_condition = ${ini_temp}
  [../]
[]

[AuxVariables]
  [./heat]
    family = MONOMIAL
    order = FIRST
    block = 'inner'
  [../]
[]

[Kernels]
  [./outer_inner_temperature_time]
    type = ADHeatConductionTimeDerivative
    variable = T
    block = 'outer inner'
    density_name = 'rho'
    specific_heat = 'cp'
  [../]
  [./outer_inner_temperature_conduction]
    type = ADHeatConduction
    variable = T
    block = 'outer inner'
    thermal_conductivity = 'k'
  [../]
  [./heat_source]
    type = HeatSource
    variable = T
    block = 'inner'
    value = 1e5
  [../]
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = FINITE
    automatic_eigenstrain_names = true
    generate_output = 'vonmises_stress'
    block = 'outer inner'
  []
[]

[Materials]
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e11
    poissons_ratio = 0.32
    block = 'outer inner'
  []
  [thermal_expansion]
    type = ComputeThermalExpansionEigenstrain
    temperature = T
    thermal_expansion_coeff = 1.0358e-5
    stress_free_temperature = 273
    eigenstrain_name = thermal_expansion
    block = 'outer inner'
  []
  [stress]
    type = ComputeFiniteStrainElasticStress
    block = 'outer inner'
  []

  [./rho_inner]
    type = ADParsedMaterial
    f_name = rho
    function = '0.0110876 * pow(9.9672e-1 + 1.179e-5 * T - 2.429e-9 * pow(T,2) + 1.219e-12 * pow(T,3),-3)'
    args = 'T'
    block = 'inner'
  [../]
  [./cp_inner]
    type = ADParsedMaterial
    f_name = cp
    function = '0.76 * ((302.27 * pow((548.68/T),2) * exp(548.68 / T)) / pow((exp(548.68 / T) - 1),2) + 2 * 8.463e-3 * T + 8.741e7 * 18531.7 * exp(-18531.7 / T) / pow(T,2)) + 0.24 * ((322.49 * pow((587.41/T),2) * exp(587.41 / T)) / pow((exp(587.41 / T) - 1),2) + 2 * 1.4679e-2 * T)'
    args = 'T'
    block = 'inner'
  [../]
  [./k_inner]
    type = ADParsedMaterial
    f_name = k
    function = '1.158/(7.5408 + 17.692 * (T / 1000) + 3.6142 * pow((T/1000),2)) + 74.105 * pow((T / 1000),-2.5) * exp(-16.35 / (T / 1000))'
    args = 'T'
    block = 'inner'
  [../]

  [./rho_outer]
    type = ADParsedMaterial
    f_name = rho
    function = '1e-6 * (7830.853 - 0.212046 * T - 1.011373e-4 * pow(T,2))'
    args = 'T'
    block = 'outer'
  [../]
  [./cp_outer]
    type = ADParsedMaterial
    f_name = cp
    function = '5863.9 - 32.563 * T + 0.072564 * pow(T,2) - 7.045375e-5 * pow(T,3) + 2.585336e-8 * pow(T,4)'
    args = 'T'
    block = 'outer'
  [../]
  [./k_outer]
    type = ADParsedMaterial
    f_name = k
    function = '0.3'
    args = 'T'
    block = 'outer'
  [../]
[]

[BCs]
  [./top_and_bottom]
    type = DirichletBC
    variable = disp_z
    boundary = 'outer_bottom outer_top inner_bottom inner_top'
    value = 0
  [../]
  [./outer_wall_T]
    type = DirichletBC
    variable = T
    value = ${ini_temp}
    boundary = 'outer_wall'
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  num_steps = 25
  dt = 0.001
  #end_time = 100

  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-10

  solve_type = 'NEWTON'
  petsc_options = '-snes_converged_reason -ksp_converged_reason -snes_linesearch_monitor'
  petsc_options_iname = '-pc_type -pc_factor_shift_type'
  petsc_options_value = 'lu       NONZERO'
  line_search = 'none'
   # petsc_options_iname = '-snes_type'
  # petsc_options_value = 'test'

  nl_max_its = 30
  l_max_its = 100
  automatic_scaling = true
[]

[Outputs]
  exodus = true
[]
```

Thanks a lot.

**Comment by hugary1995** [comment, marked as answer]
For the difference you see with different processors, that's very likely due to rigid body modes in the xy plane. You can still solve the system because you are using mumps with multiple processors plus a nonzero diagonal shift, but you lose the uniqueness of the solution.

**Reply by hugary1995** [reply]
Same reason for the asymmetry. There are two ways to resolve both issues:
1. Remove the rigid body modes by fixing a point in the domain. Any point should do.
2. Add an inertia term in the momentum balance.

**Reply by js-jixu** [reply]
Do you mean that I should set boundary conditions on a point in the domain? I set new bundary conditions like this;
```
  [./bottom_point_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'bottom_point'
    value = 0
  [../]
  [./bottom_point_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom_point'
    value = 0
  [../]
  [./bottom_point_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'bottom_point'
    value = 0
  [../]
```

The `bottom_point` is the point in the middle of the bottom of the cylinder with coordinates (0, 0, 0). I think if the input file is correct, the result will be the same whether it is a quad-core or an octa-core calculation. So I computed the input file with the new boundary conditions with four and eight cores respectively. . But the results calculated by different numbers of cores are still different and  asymmetry. Here are outputs screenshots and files:
![4cores](https://user-images.githubusercontent.com/99813361/185639464-b26b4415-5a03-495d-b22e-6f838c3a0ab4.png)
![8cores](https://user-images.githubusercontent.com/99813361/185639479-b39198bb-d8d8-4ff4-b884-bf21de4a539e.png)
[files.zip](https://github.com/idaholab/moose/files/9382849/files.zip)

If I want to set the model to have constraints at the top and bottom. In antoher word, the model cannot expand in the axial direction, only in the radial direction, how should I set the boundary conditions?

**Reply by hugary1995** [reply]
> Remove the rigid body modes by fixing a point in the domain. Any point should do.

Uhh I guess I didn't make it clear, my bad. The way you are fixing the point removes the in-plane translational motion, but the body could still rotate. You'd want to pick two points, fix x and y at the first point, and fix x or y at the second point. You'd want to keep the original BC in the z direction.

**Reply by js-jixu** [reply]
I've set boundary conditions on `top_point` and `bottom_point`, which are the midpoints of the top and bottom of the model, the new BCs block is here:
```
[BCs]
  [./bottom_point_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom_point'
    value = 0
  [../]
  [./bottom_point_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'bottom_point'
    value = 0
  [../]
  [./top_point_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'top_point'
    value = 0
  [../]
  [./top_and_bottom_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'outer_bottom outer_top inner_bottom inner_top'
    value = 0
  [../]
  [./outer_wall_T]
    type = DirichletBC
    variable = T
    value = ${ini_temp}
    boundary = 'outer_wall'
  [../]
[]
```

But different numbers of cores still compute different results. The following are the results of the calculations for quad-core and octa-core:
![4cores](https://user-images.githubusercontent.com/99813361/185719609-d592194d-039a-4621-b2d2-add282cfc673.png)
![8cores](https://user-images.githubusercontent.com/99813361/185719612-e8a76048-bc62-45c0-b45e-f5e756236e9f.png)
[new_files.zip](https://github.com/idaholab/moose/files/9385843/new_files.zip)

The `new_files` folder has a mesh file which has a coarser mesh in the axial direction.

**Reply by hugary1995** [reply]
Think about it: the "top_point" and the "bottom_point" you picked happen to align with the center axis. So the cylinder could still rotate (about the center axis).

Try this: say the radius of the cylinder is R, and the bottom of the cylinder has origin (0, 0, 0), then pick point 1 at (0, 0, 0) and point 2 at (R, 0, 0). Fix the x and y displacements at point 1, and fix the y displacement at point 2.

**Reply by js-jixu** [reply]
Amazing idea! You are totally correct, Gary! I've got reasonable results after taking your advice. Sincere thanks to your patience and wisdom.

**Reply by hugary1995** [reply]
So I just tested with your coarse mesh, and the results are consistent using different number of processors. My mesh block and BCs are
```python
[Mesh]
  [fmg]
    type = FileMeshGenerator
    file = '2layers_3d_4parts_coarse.msh'
  []
  [pin]
    type = ExtraNodesetGenerator
    input = fmg
    new_boundary = 'pin'
    nodes = 4566
  []
[]

[BCs]
  [bottom_point_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom_point'
    value = 0
  []
  [bottom_point_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'bottom_point'
    value = 0
  []
  [pin_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'pin'
    value = 0
  []
  [bottom_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'outer_bottom inner_bottom'
    value = 0
  []
  [outer_wall_T]
    type = DirichletBC
    variable = T
    value = ${ini_temp}
    boundary = 'outer_wall'
  []
[]
```

**Reply by js-jixu** [reply]
Yeah, the results is absolutely correct! Thanks again!

**Reply by hugary1995** [reply]
If you want the body to expand/shrink in the z-direction (due to thermal expansion for example), you probably don't want to constrain both top and bottom. Just constrain the z displacements on the bottom surface should suffice. You may want to constrain the top surface to have the same z-displacement, in which case you could use an `EqualValueBoundaryConstraint` on the top surface.

**Reply by js-jixu** [reply]
Okay, I'll learn to use this object carefully. Your advice really helps a lot to me!
